src/metrics/pocket_area/adaptive_pocket_area.py

```diff
@@ -0,0 +1,108 @@
+    3. Filter out all players whose distance from the passer is greater than that distance
+    4. If the filtered list has more than two players, calculate the convex hull of their coordinates and return the area
+    5. If there are exactly two, calculate the area of a circle with that distance as the radius, then multiply that area by a proportion (let's say 0.25 for now) to represent the cone in front of the passer
+    6. If there are fewer than two players in the filtered list raise an exception because something has gone wrong
+
+"""
+
+
+def find_closest_player_s(
+    point: Dict,
+    blockers: List[Dict],
+    rushers: List[Dict],
+    rusher_difference: float,
+):
+    """Returns the closest players to the given point and the distance
+
+    rusher_difference : acceptable difference parameter from the closest rusher
+    and any other player that should be considered to make the pocket
+    """
+    if not rushers:
+        raise InvalidPocketError("No rusher to guage pocket.")
+
+    closest_players: List[Dict] = []
+    closest_distance = float("inf")
+
+    # Find the closest rusher's distance
+    for player in rushers:
+        d = get_distance(point, player)
+        if d < (closest_distance) or closest_distance == float("inf"):
+            closest_distance = d
+
+    players = blockers + rushers
+
+    # Determine whether each rusher is within the rusher difference error
+    for player in players:
+        d = get_distance(point, player)
+        if d <= (closest_distance + rusher_difference):
+            closest_players.append(player)
+
+    return closest_players, closest_distance
+
+
+# Corvex Hull function with blockers and rushers combined
+def get_convexHull_area(frame: List[Dict]) -> PocketArea:
+    """
+    Estimates the pocket area as the convex hull of all the pass blockers on the field
+    """
+
+    pocket_players = frame
+    area, vertices = get_convex_hull(pocket_players)
+    metadata = PocketAreaMetadata(vertices=vertices)
+    return PocketArea(area, metadata)
+
+
+def calculate_adaptive_pocket_area(frame: pd.DataFrame) -> PocketArea:
+
+    passer, blockers, rushers = split_records_by_role(frame)
+    if not rushers:
+        raise InvalidPocketError("No rushers in frame to make pocket.")
+    if not passer:
+        raise InvalidPocketError("No passer in frame.")
+
+    # Consider all pass rushers and the passer
+    pocket_players = rushers + blockers + [passer]
+    # use helper function in this file to get all the rushers that have a distance 3 yards
+    # within the distance from the closest rusher to the quarterback
+    closest_rushers = find_closest_player_s(passer, blockers, rushers, 1.0)
+
+    # adjusted pocket will get all the pass rushers that make a valid pocket along with the qb
+    adjusted_pocket = closest_rushers[0] + [passer]
+
+    # If there is 2 or more valid pass rushers, get the corvex hull of the rushers and qb
+    if len(closest_rushers[0]) >= 2:
```

Comment:
This should be greater than two. If we only have two points, we can't calculate the convex hull.

Reply:
closest_rushers[0] (now closest_lineman in the next push) only contain the closest lineman, the adapted_pocket also contains the passer so there will be 3 points.

Reply:
Ah, I see, that makes sense. Thanks for explaining!